[SYCL][ESIMD] Add tests on simd increment and decrement operators

vasilytric opened the pull request.

vasilytric committed:
[SYCL][ESIMD] Add tests on simd increment and decrement operators

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,254 @@
+
+// Descriptor class for the case of calling constructor in initializer context.
+struct pre_decrement {
+  static std::string get_operator_type() { return " pre decrement"; }
```
```suggestion
  static std::string get_operator_type() { return "pre decrement"; }
```
Looks like unintended whitespace left for all 4 test cases

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/827/commits/07503714b6bf2d29f8318597e51142bec03a1fb8#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eL28)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,254 @@
+
+// Descriptor class for the case of calling constructor in initializer context.
+struct post_decrement {
+  static std::string get_operator_type() { return " post decrement"; }
```
Looks like `type` suffix is misleading here.

I would suggest either to use the same name as for all other test cases
```suggestion
  static std::string get_description() { return "post decrement"; }
```
or to use `named_type_pack` alongside with
```suggestion
  static std::string to_string() { return "post decrement"; }
```

vasilytric replied:
this comment was applied in this commit: d0f0c3a641f23798485b102a95bc802b53957700

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,254 @@
+  static constexpr bool is_increment() { return true; }
+};
+
+template <typename DataT, int NumElems, typename ContextT>
```
```suggestion
template <typename DataT, int NumElems, typename TestCaseT>
```

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/827/commits/07503714b6bf2d29f8318597e51142bec03a1fb8#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eL114)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,254 @@
+    result_simd.copy_to(result_simd_out);
+  }
+
+  template <typename DataT> static DataT apply_operator(DataT val) {
+    return ++val;
+  }
```
```suggestion
  template <typename DataT> static DataT apply_operator(DataT& val) {
    return ++val;
  }
```
By doing so you would be able:

- to remove `is_increment()` entirely
- to simplify "expected values" logic within `run_verification`
- to have exactly the post-increment/decrement operator within `apply_operator` for post-increment/decrement test case

For example:
```
DataT expected_source_value = shared_ref_data[i];
const DataT expected_return_value = TestCaseT::apply_operator(expected_source_value );
```

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/827/commits/07503714b6bf2d29f8318597e51142bec03a1fb8#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eL205-R210)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,254 @@
+
+      expected_value = TestCaseT::apply_operator(expected_value);
+      passed &= verify_result(i, expected_value, result_simd_out[i],
+                              "return value for", data_type);
```
Would you mind to provide an example of failure output for both `verify_result` calls?
Anyway, I would suggest to use something regarding "specific check" instead of `source_or_destination_simd` and `context` variable names.

vasilytric replied:
We will have the same output like this: Failed for simd<sycl::half, 8>, retrieved: -nan, expected: -nan(ind), at index: 1 for simd as a return value for pre increment operator.

> Anyway, I would suggest to use something regarding "specific check" instead of source_or_destination_simd and context variable names.

I will think about the new names

vasilytric replied:
choose the right name is very hard thing in this case, I've tried to apply this comment [here](https://github.com/intel/llvm-test-suite/pull/827/commits/07503714b6bf2d29f8318597e51142bec03a1fb8#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eL247-L248)

yuriykoch replied:
Agree, I would suggest something regarding `check_description` though. The `simd_type` is more like the `simd<DataT, NumElems>` itself.

vasilytric replied:
@yuriykoch which name you will suggest instead of current name?

yuriykoch replied:
What about the following instead of `check_description` or `specific_check`:
```
      passed &= verify_result(i, expected_source_value, source_simd_out[i],
                              "unexpected argument modification", data_type);
      passed &= verify_result(i, expected_return_value, result_simd_out[i],
                              "unexpected return value", data_type);

....

    log_msg += ", at index: " + std::to_string(m_index);
    log_msg += "for " + TestCaseT::get_description() + " operator: ";
    log_msg += m_error_details;
```

vasilytric replied:
new error messages will be looks like:
```
Failed for simd<float, 32>, retrieved: 4.100000, expected: 4.100000, at index: 31 for post increment operator: unexpected argument modification
Failed for simd<float, 32>, retrieved: 3.100000, expected: 3.100000, at index: 31 for post increment operator: unexpected return value
```
I think that these is ok messages

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment_fp_extra.cpp:
```diff
@@ -0,0 +1,51 @@
+                    esimd_test::createExceptionHandler());
+
+  bool passed = true;
+#ifdef RUN_TEST_WITH_HALF
```
```suggestion
#ifdef ESIMD_TESTS_RUN_WITH_HALF
```

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/827/commits/07503714b6bf2d29f8318597e51142bec03a1fb8#diff-8bfd87b59402abfe7e7ca599205da194d65cb2a3f3dbf151a6687b21baf62758L35-L36)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,254 @@
+  }
+
+  template <typename DataT> static DataT apply_operator(DataT val) {
+    return val;
```
```suggestion
    return val--;
```

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/827/commits/07503714b6bf2d29f8318597e51142bec03a1fb8#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eR64)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,254 @@
+    result_simd.copy_to(result_simd_out);
+  }
+
+  template <typename DataT> static DataT apply_operator(DataT val) {
+    return val;
+  }
```
```suggestion
  template <typename DataT> static DataT apply_operator(DataT& val) {
    return val++;
  }
```
That's what user would expect from calling `post_increment::apply_operator()` API

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/827/commits/07503714b6bf2d29f8318597e51142bec03a1fb8#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eR108)

vasilytric committed:
[SYCL][ESIMD] Some structural changes

- change verification logic
- rename variable

vasilytric committed:
[SYCL][ESIMD] Add possible to run fp accuracy tests from existed run_test

It provides use an existed common code without copy pasting

vasilytric committed:
Merge branch 'intel' into operator_incr_decr

vasilytric committed:
[SYCL][ESIMD] Replace dim with size

vasilytric committed:
[SYCL][ESIMD] Replace get_operator_type with get_description

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,283 @@
+    bool passed = true;
+    std::vector<DataT> ref_data;
+
+    if constexpr (!IsAccuracyTestT::value) {
```
Looks like it would be better to move out L#160 - L#196 to a separate function. These lines of code have a clear intent: to generate input data for the test.

As the next step of refactoring, I would suggest to use separate classes for different `IsAccuracyTestT` values.
You are actually using different logic of input values generation based on either `is_fp_accuracy_test` or `is_base_test`. So why not to go a bit further and do something like this:
```
struct base_test {
  template <typename DataT, int NumElems, typename TestCaseT>
  std::vector<DataT> generate_ref_data();
};
struct fp_accuracy_test {
  template <typename DataT, int NumElems, typename TestCaseT>
  std::vector<DataT> generate_ref_data();
};
```

vasilytric replied:
`base_test` and `fp_accuracy_test` can be found  [here](https://github.com/intel/llvm-test-suite/pull/827/commits/7ff74964cc8f28e39cce66778060f321038d13cc#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eR149-R197)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,283 @@
+      static const DataT min = value<DataT>::lowest();
+      static const DataT denorm_min = value<DataT>::denorm_min();
+      static const DataT max = value<DataT>::max();
+      static const DataT zero_point_one = static_cast<DataT>(0.1);
```
```suggestion
      static const DataT unexact = static_cast<DataT>(0.1);
```
Please, avoid naming any variable based on its exact value. Generalize the code

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/827/commits/7ff74964cc8f28e39cce66778060f321038d13cc#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eR172)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,283 @@
+      static const DataT denorm_min = value<DataT>::denorm_min();
+      static const DataT max = value<DataT>::max();
+      static const DataT zero_point_one = static_cast<DataT>(0.1);
+      static const DataT one_point_zero = 1.0;
```
What's the intent behind?

vasilytric replied:
`one_point_zero ` has been removed

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,283 @@
+        ref_data.reserve((NumElems > 1) ? NumElems : 6);
+        ref_data.insert(ref_data.end(),
+                        {zero_point_one, denorm_min, -denorm_min,
+                         -one_point_zero + value<DataT>::pos_ulp(42U),
+                         min + value<DataT>::pos_ulp(42U),
+                         (max - 1) + value<DataT>::neg_ulp(42U)});
```
Here and elsewhere: looks like misuse of `neg_ulp` a bit. We would have a UB for (max - 1 + 42 - 1ULP). What about something like this?
```suggestion
                         value<DataT>::pos_ulp(-one_point_zero),
                         value<DataT>::pos_ulp(min),
                         value<DataT>::neg_ulp(max - 1)});
```

vasilytric replied:
this comment applied [here](https://github.com/intel/llvm-test-suite/pull/827/commits/7ff74964cc8f28e39cce66778060f321038d13cc#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eR174-R192)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,283 @@
+                         min + value<DataT>::pos_ulp(42U),
+                         (max - 1) + value<DataT>::neg_ulp(42U)});
+        for (size_t i = ref_data.size(); i < NumElems; ++i) {
+          ref_data.push_back(0.1 * i);
```
```suggestion
          ref_data.push_back(unexact * i);
```
Here and below

vasilytric replied:
this comment applied [here](https://github.com/intel/llvm-test-suite/pull/827/commits/7ff74964cc8f28e39cce66778060f321038d13cc#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eR172)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,283 @@
+
+    queue.submit([&](sycl::handler &cgh) {
+      const DataT *const ref = shared_ref_data.data();
+      DataT *source_simd_data_storage = source_simd_out.data();
+      DataT *result_simd_data_storage = result_simd_out.data();
```
Nit, just as suggestion:
```suggestion
      DataT *source_simd_data_ptr = source_simd_out.data();
      DataT *result_simd_data_ptr = result_simd_out.data();
```
it's more like `source_simd_out` is a storage for data, and these ones are the USM pointers to access the storage

vasilytric replied:
this comment applied [here](https://github.com/intel/llvm-test-suite/pull/827/commits/7ff74964cc8f28e39cce66778060f321038d13cc#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eL228-R244)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,283 @@
+  }
+
+  bool verify_result(size_t i, DataT expected, DataT retrieved,
+                     std::string &&simd_type, const std::string &data_type) {
```
```suggestion
                     const std::string &simd_type, const std::string &data_type) {
```
There is no profit here:

- no std::move
- no need in r-value restriction
- we need to copy this value in `IncrementAndDecrementTestDescription` anyway

vasilytric replied:
this comment applied [here](https://github.com/intel/llvm-test-suite/pull/827/commits/7ff74964cc8f28e39cce66778060f321038d13cc#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eL122-R122)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment_accuracy_core.cpp:
```diff
@@ -0,0 +1,45 @@
+
+  bool passed = true;
+
+  const auto u_types = get_tested_types<tested_types::fp>();
```
It's a bit unclear for me what does `u` in `u_types` mean
```suggestion
  const auto fp_types = get_tested_types<tested_types::fp>();
```

vasilytric replied:
this comment applied [here](https://github.com/intel/llvm-test-suite/pull/827/commits/7ff74964cc8f28e39cce66778060f321038d13cc#diff-009d4498abf01fa40a27db546175221887f54a04b58f5db8e4eb7fbbf1a1af1fL33-R33)

vasilytric committed:
[SYCL][ESIMD] Small updates

- Change log message
- Small refactoring
- variable renaming

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,299 @@
+      for (size_t i = ref_data.size(); i < NumElems; ++i) {
+        ref_data.push_back(inexact * i);
+      }
+    } else if constexpr (!TestCaseT::is_increment()) {
```
```suggestion
    } else {
```

vasilytric replied:
fixed [here](https://github.com/intel/llvm-test-suite/pull/827/commits/7dddeb4523bc5d5bf8bb785d6f17a1c6e2461de4#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eL184)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,299 @@
+                       value<DataT>::neg_ulp(static_cast<DataT>(-1.0)),
+                       value<DataT>::neg_ulp(max),
+                       value<DataT>::pos_ulp(min + 1)});
+      for (size_t i = ref_data.size(); i < NumElems; ++i) {
+        ref_data.push_back(inexact * i);
+      }
```
Maybe we might move the same blocks out of if-else branches?
```
if () {
} else {
}
// Fill the reference data to be no less than simd vector length
...
return ref_dat;
```

vasilytric replied:
fixed [here](https://github.com/intel/llvm-test-suite/pull/827/commits/7dddeb4523bc5d5bf8bb785d6f17a1c6e2461de4#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eL181-R191)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp:
```diff
@@ -0,0 +1,299 @@
+        log::note(log_msg);
+      }
+    }
+    if (are_bitwise_equal(expected, retrieved)) {
+      passed = false;
```
```suggestion
    if (!are_bitwise_equal(expected, retrieved)) {
      passed = false;
```
Is it a misprint?

vasilytric replied:
Yes, fixed [here](https://github.com/intel/llvm-test-suite/pull/827/commits/7dddeb4523bc5d5bf8bb785d6f17a1c6e2461de4#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eR284)

vasilytric committed:
[SYCL][ESIMD] Update logic statements

- Remove unnecessary if else
- Remove duplicated for cycle
- Fix logic statement in "if" condition
- Remove unused aliases

vasilytric committed:
[SYCL][ESIMD] Use actual statements for fp_accuracy_test and base_test

vasilytric commented:
@sndmitriev Would you mind to review this PR?

sndmitriev commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment_accuracy_core.cpp:
```diff
@@ -0,0 +1,45 @@
+//==------- operator_logical_not.cpp  - DPC++ ESIMD on-device test ---------==//
```
nit: file name is not correct.

vasilytric replied:
this comment has been applied in this commit: 75b472db1fe0777cd84eee7aca25852f5a44d436

sndmitriev commented on SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment_accuracy_fp_extra.cpp:
```diff
@@ -0,0 +1,53 @@
+//==------- operator_logical_not.cpp  - DPC++ ESIMD on-device test ---------==//
```
and here as well.

vasilytric replied:
this comment has been applied in this commit: 75b472db1fe0777cd84eee7aca25852f5a44d436

vasilytric committed:
[SYCL][ESIMD] Use correct file hames in header comment

vasilytric commented:
@kbobrovs Would you mind to merge this PR?
